actors/builtin/miner/miner_state.go

```diff
@@ -267,64 +306,430 @@ func (st *State) ForEachSector(store adt.Store, f func(*SectorOnChainInfo)) erro
-		if !fault {
-			sectorInfos = append(sectorInfos, abi.SectorInfo{
-				SealedCID:       ssinfo.Info.SealedCID,
-				SectorNumber:    ssinfo.Info.SectorNumber,
-				RegisteredProof: ssinfo.Info.RegisteredProof,
-			})
+	// Load the sector infos, masking out fault sectors with a good one.
+	var sectorInfos []*SectorOnChainInfo
+	err = abi.BitFieldForEach(sectors, func(i uint64) error {
+		sector := standInInfo
+		faulty := faultSet[i]
+		if !faulty {
+			sectorOnChain, found, err = st.GetSector(store, abi.SectorNumber(i))
+			if err != nil {
+				return fmt.Errorf("failed to load sector %d: %v", i, err)
+			} else if !found {
+				return fmt.Errorf("can't find sector %d", i)
+			}
+			sector = sectorOnChain
 		}
+		sectorInfos = append(sectorInfos, sector)
 		return nil
 	})
+	return sectorInfos, err
+}
+
+// Adds partition numbers to the set of PoSt submissions
+// Precondition: no partitions are yet in PoSt submissions
+func (st *State) AddPoStSubmissions(partitionNos []uint64) error {
+	mp, err := st.PostSubmissions.AllMap(WPoStPeriodDeadlines)
+	if err != nil {
+		return err
+	}
+
+	for _, no := range partitionNos {
+		if exist := mp[no]; exist {
+			return xerrors.Errorf("duplicate PoSt submission for partition %d", no)
+		}
+
+		st.PostSubmissions.Set(no)
+	}
+
+	return nil
+}
+
+// Removes all PoSt submissions
+func (st *State) ClearPoStSubmissions() error {
+	st.PostSubmissions = abi.NewBitField()
+	return nil
+}
+
+//
+// Window PoSt Deadlines
+//
+
+// Returns an array of the count of sectors in each deadline.
+func (d *Deadlines) CountDeadlines() ([WPoStPeriodDeadlines]uint64, error) {
+	// TODO REVIEW: bitfield.Count() is not constant time, consider storing counts explicitly
+	var out [WPoStPeriodDeadlines]uint64
+
+	for i, dl := range d.Due {
+		var err error
+		out[i], err = dl.Count()
+		if err != nil {
+			return [WPoStPeriodDeadlines]uint64{}, err
+		}
+	}
+
+	return out, nil
+}
+
+// Adds sector numbers to a deadline.
+func (d *Deadlines) AddToDeadline(deadline int, newSectors ...abi.SectorNumber) (err error) {
+	ns := abi.NewBitField()
+	for _, sector := range newSectors {
+		ns.Set(uint64(sector))
+	}
+
+	d.Due[deadline], err = abi.BitFieldMerge(d.Due[deadline], ns)
+	return err
+}
 
+// Removes sector numbers from all deadlines.
+func (d *Deadlines) RemoveFromAllDeadlines(sectorNos abi.BitField) error {
+	ns, err := sectorNos.All(MaxFaultsCount) // TODO: MaxSeals?
 	if err != nil {
-		return sectorInfos, errors.Wrapf(err, "failed to traverse sectors for proving set: %v", err)
+		return err
 	}
 
-	return sectorInfos, nil
+	for i := range d.Due {
+		// TODO: bitfield.Sub
+
+		for _, n := range ns {
+			d.Due[i].Unset(n)
+		}
+	}
+	return nil
 }
 
+// Computes the first partition index and number of sectors for a deadline.
+// Partitions are numbered globally for the miner, not per-deadline.
+func (d *Deadlines) PartitionsForDeadline(deadlineIndex uint64) (firstPartition, sectorCount uint64, _ error) {
```

Comment:
@anorth What is meant by "first partition index"?

Reply:
It's the index of the first (lowest) partition number due at the requested deadline. The partitions are indexed from 0 by enumerating those in the first deadline, then the second, etc for all the deadlines.